### PR TITLE
gitserver: Set clone status after attempting to clone repo

### PR DESCRIPTION
**Pull request by ryanslade**
We now set the status of a repo to `cloning` when we begin cloning it, unless it
has already been cloned. Once we complete the cloning process we again set the
state in the db to either `cloned` or `not_cloned`.

### CODE REVIEW

File: cmd/gitserver/server/server.go [hunk 1]
```diff
@@ -1214,6 +1215,23 @@ func (s *Server) cloneRepo(ctx context.Context, repo api.RepoName, opts *cloneOp
...
+				// No need to stop cloning
+				log15.Warn("Setting clone status in DB", "error", err)
+			}
+		}
+		defer func() {
+			status := types.CloneStatusCloned
+			if !repoCloned(dir) {
+				status = types.CloneStatusNotCloned
+			}
+			if err := s.setCloneStatus(ctx, repo, status); err != nil {
```
Comment (keegancsmith) [inline cmd/gitserver/server/server.go:1230]: if we fail due to timeout this context will be canceled, so you won't be able to update teh status to notcloned.

File: cmd/gitserver/server/server.go [hunk 2]
```diff
@@ -1214,6 +1215,23 @@ func (s *Server) cloneRepo(ctx context.Context, repo api.RepoName, opts *cloneOp
...
+				log15.Warn("Setting clone status in DB", "error", err)
+			}
+		}()
```
Comment (keegancsmith) [inline cmd/gitserver/server/server.go:1233]: this feels like a lot of code in an already biggish function. Is there a way to make the impact on the function smaller? eg factor out the bulk of the reporting into a helper function(s).

File: cmd/gitserver/server/server.go [hunk 3]
```diff
@@ -1214,6 +1215,23 @@ func (s *Server) cloneRepo(ctx context.Context, repo api.RepoName, opts *cloneOp
 		tmpPath = filepath.Join(tmpPath, ".git")
 		tmp := GitDir(tmpPath)
 
+		// It may already be cloned
+		if !repoCloned(dir) {
+			if err := s.setCloneStatus(ctx, repo, types.CloneStatusCloning); err != nil {
```
Comment (keegancsmith) [inline cmd/gitserver/server/server.go:1220]: nit: if the callsite never changes behaviour due to error, why not not return error and instead log internally?

**Review by keegancsmith — APPROVED** [review]

File: cmd/gitserver/server/server.go (hunk 3 above)
Comment (ryanslade) [reply cmd/gitserver/server/server.go:1220]: I don't want to hide the error and I'd rather the call site decides whether failing to save the error is fatal or not. Currently the call sites just log but I don't want to assume all future callers want that too.

**Comment by ryanslade** [comment]
> General question: what happens if the service halt and leaving the state to be `cloning`, does it hurt?

It depends how long the service is down. If it comes back up quickly then the `SyncRepoState` method will run and get things back in line. If it's down for a while then it'll stay set as `cloning` until it's back. I don't think it's too much of a problem though.

**Pull request merged**